File: r2r/cli/commands/server_operations.py
import json
import os
import platform
import subprocess
import sys

import click
from dotenv import load_dotenv

from r2r.cli.command_group import cli
from r2r.cli.utils.docker_utils import (
    bring_down_docker_compose,
    remove_r2r_network,
    run_docker_serve,
)
from r2r.cli.utils.timer import timer
from r2r.main.execution import R2RExecutionWrapper


@cli.command()
@click.option(
    "--volumes",
    is_flag=True,
    help="Remove named volumes declared in the `volumes` section of the Compose file",
)
@click.option(
    "--remove-orphans",
    is_flag=True,
    help="Remove containers for services not defined in the Compose file",
)
@click.option("--project-name", default="r2r", help="Project name for Docker")
@click.pass_context
def docker_down(ctx, volumes, remove_orphans, project_name):
    """Bring down the Docker Compose setup and attempt to remove the network if necessary."""
    result = bring_down_docker_compose(project_name, volumes, remove_orphans)

    if result != 0:
        click.echo(
            "An error occurred while bringing down the Docker Compose setup. Attempting to remove the network..."
        )
        remove_r2r_network()
    else:
        click.echo("Docker Compose setup has been successfully brought down.")


@cli.command()
def generate_report():
    """Generate a system report including R2R version, Docker info, and OS details."""

    # Get R2R version
    from importlib.metadata import version

    report = {"r2r_version": version("r2r")}

    # Get Docker info
    try:
        # Get running containers
        docker_ps_output = subprocess.check_output(
            ["docker", "ps", "--format", "{{.ID}}\t{{.Names}}\t{{.Status}}"],
            text=True,
        ).decode("utf-8")
        report["docker_ps"] = [
            dict(zip(["id", "name", "status"], line.split("\t")))
            for line in docker_ps_output.strip().split("\n")
            if line
        ]

        # Get running subnets in Docker
        docker_network_output = subprocess.check_output(
            ["docker", "network", "ls", "--format", "{{.ID}}\t{{.Name}}"],
            text=True,
        ).decode("utf-8")
        networks = [
            dict(zip(["id", "name"], line.split("\t")))
            for line in docker_network_output.strip().split("\n")
            if line
        ]

        report["docker_subnets"] = []
        for network in networks:
            inspect_output = subprocess.check_output(
                [
                    "docker",
                    "network",
                    "inspect",
                    network["id"],
                    "--format",
                    "{{range .IPAM.Config}}{{.Subnet}}{{end}}",
                ],
                text=True,
            ).decode("utf-8")
            if subnet := inspect_output.strip():
                network["subnet"] = subnet
                report["docker_subnets"].append(network)

    except subprocess.CalledProcessError as e:
        report["docker_error"] = f"Error running Docker command: {e}"
    except FileNotFoundError:
        report["docker_error"] = (
            "Docker command not found. Is Docker installed and in PATH?"
        )

    # Get OS information
    report["os_info"] = {
        "system": platform.system(),
        "release": platform.release(),
        "version": platform.version(),
        "machine": platform.machine(),
        "processor": platform.processor(),
    }

    click.echo("System Report:")
    click.echo(json.dumps(report, indent=2))


@cli.command()
@click.pass_obj
def health(obj):
    """Check the health of the server."""
    with timer():
        response = obj.health()

    click.echo(response)


@cli.command()
@click.option("--host", default="0.0.0.0", help="Host to run the server on")
@click.option("--port", default=8000, help="Port to run the server on")
@click.option("--docker", is_flag=True, help="Run using Docker")
@click.option(
    "--exclude-neo4j", is_flag=True, help="Exclude Neo4j from Docker setup"
)
@click.option(
    "--exclude-ollama", is_flag=True, help="Exclude Ollama from Docker setup"
)
@click.option(
    "--exclude-postgres",
    is_flag=True,
    help="Exclude Postgres from Docker setup",
)
@click.option("--project-name", default="r2r", help="Project name for Docker")
@click.option(
    "--config-path",
    type=click.Path(exists=True),
    help="Path to the configuration file",
)
@click.option("--image", help="Docker image to use")
@click.pass_obj
def serve(
    obj,
    host,
    port,
    docker,
    exclude_neo4j,
    exclude_ollama,
    exclude_postgres,
    project_name,
    config_path,
    image,
):
    """Start the R2R server."""
    load_dotenv()

    if config_path:
        config_path = os.path.abspath(config_path)

        # For Windows, convert backslashes to forward slashes and prepend /host_mnt/
        if platform.system() == "Windows":
            config_path = "/host_mnt/" + config_path.replace(
                "\\", "/"
            ).replace(":", "")

        obj["config_path"] = config_path

    if docker:
        run_docker_serve(
            obj,
            host,
            port,
            exclude_neo4j,
            exclude_ollama,
            exclude_postgres,
            project_name,
            config_path,
            image,
        )
    else:
        run_local_serve(obj, host, port)


def run_local_serve(obj, host, port):
    wrapper = R2RExecutionWrapper(**obj, client_mode=False)
    wrapper.serve(host, port)


@cli.command()
def update():
    """Update the R2R package to the latest version."""
    try:
        cmd = [sys.executable, "-m", "pip", "install", "--upgrade", "r2r"]

        click.echo("Updating R2R...")
        result = subprocess.run(
            cmd, check=True, capture_output=True, text=True
        )
        click.echo(result.stdout)
        click.echo("R2R has been successfully updated.")
    except subprocess.CalledProcessError as e:
        click.echo(f"An error occurred while updating R2R: {e}")
        click.echo(e.stderr)
    except Exception as e:
        click.echo(f"An unexpected error occurred: {e}")


@cli.command()
def version():
    """Print the version of R2R."""
    from importlib.metadata import version

    click.echo(version("r2r"))

File: go/cli/server_operations.go
package main

import (
	"encoding/json"
	"fmt"
	"os/exec"
	"runtime"
	"runtime/debug"

	"github.com/spf13/cobra"
	"github.com/yourusername/r2r_go_sdk/pkg/sdk"
)

func init() {
	rootCmd.AddCommand(dockerDownCmd)
	rootCmd.AddCommand(healthCmd)
	rootCmd.AddCommand(generateReportCmd)

	dockerDownCmd.Flags().Bool("volumes", false, "Remove named volumes")
	dockerDownCmd.Flags().Bool("remove-orphans", false, "Remove containers for services not defined in the Compose file")
	dockerDownCmd.Flags().String("project-name", "r2r", "Project name for Docker")
}

// TODO: Implement the actual Docker Compose down logic
var dockerDownCmd = &cobra.Command{
	Use:   "docker-down",
	Short: "Bring down the Docker Compose setup",
	Run: func(cmd *cobra.Command, args []string) {
		volumes, _ := cmd.Flags().GetBool("volumes")
		removeOrphans, _ := cmd.Flags().GetBool("remove-orphans")
		projectName, _ := cmd.Flags().GetString("project-name")

		fmt.Printf("Bringing down Docker Compose setup (Project: %s, Volumes: %v, Remove Orphans: %v)\n", projectName, volumes, removeOrphans)
	},
}

// TODO: Implement the actual system report generation logic
var generateReportCmd = &cobra.Command{
	Use:   "generate-report",
	Short: "Generate a system report",
	Run: func(cmd *cobra.Command, args []string) {
		report := make(map[string]interface{})

		// Get R2R version
		// TODO: Implement the actual version retrieval logic
		report["r2r_version"] = "1.0.0"

		// Get Docker info
		dockerVerison, _ := exec.Command("docker", "--version").Output()
		report["docker_version"] = string(dockerVerison)

		dockerPs, _ := exec.Command("docker", "ps", "--format", "{{.ID}}\t{{.Names}}\t{{.Status}}").Output()
		report["docker_info"] = string(dockerPs)

		dockerNetworks, _ := exec.Command("docker", "network", "ls", "--format", "{{.ID}}\t{{.Name}}\t{{.Driver}}").Output()
		report["docker_networks"] = string(dockerNetworks)

		dockerSubnets, _ := exec.Command("docker", "network", "inspect", "bridge", "--format", "{{.IPAM.Config}}").Output()
		report["docker_subnets"] = string(dockerSubnets)

		// Get OS information
		report["os_info"] = map[string]string{
			"system":    runtime.GOOS,
			"version":   runtime.Version(),
			"machine":   runtime.GOARCH,
			"processor": runtime.GOARCH,
		}

		jsonReport, _ := json.MarshalIndent(report, "", "  ")
		fmt.Println(string(jsonReport))
	},
}

var healthCmd = &cobra.Command{
	Use:   "health",
	Short: "Check the health of the server",
	Run: withTimer(func(cmd *cobra.Command, args []string) {
		client := sdk.NewClient("http://localhost:8000/v1", sdk.LogConfig{Verbose: true})
		response, err := client.Health()
		if err != nil {
			fmt.Printf("Error checking health: %v\n", err)
			return
		}
		fmt.Println(response)
	}),
}

// TODO: r2r serve command

// TODO: update

var versionCmd = &cobra.Command{
	Use:   "version",
	Short: "Print the version of R2R",
	Run: func(cmd *cobra.Command, args []string) {
		info, ok := debug.ReadBuildInfo()
		if !ok {
			fmt.Println("No version information available")
			return
		}

		for _, dep := range info.Deps {
			if dep.Path == "github.com/yourusername/r2r_go_sdk" {
				fmt.Printf("R2R version: %s\n", dep.Version)
				return
			}
		}

		fmt.Println("R2R version information not found")
	},
}
